File: Java/Visitor.java
import com.sun.jdi.IntegerType;
import java.awt.geom.FlatteningPathIterator;
import java.util.HashMap;
import java.util.Map;

public class EvalVisitor extends LabeledExprBaseVisitor<Double> {
    Map<String, Double> memory = new HashMap<>();

    @Override
    public Double visitAssign(LabeledExprParser.AssignContext ctx) {
        String id = ctx.ID().getText();
        Double value = visit(ctx.expr());
        memory.put(id, value);
        return value;
    }

    @Override
    public Double visitPrintExpr(LabeledExprParser.PrintExprContext ctx) {
        Double value = visit(ctx.expr());
        System.out.println(value);
        return 0.0;
    }
    @Override
    public Double visitDouble(LabeledExprParser.DoubleContext ctx) {
        return Double.valueOf(ctx.DOUBLE().getText());
    }
    @Override
    public Double visitUnaryMinus(LabeledExprParser.UnaryMinusContext ctx) {
        return -visit(ctx.expr());
    }

    @Override
    public Double visitUnaryPlus(LabeledExprParser.UnaryPlusContext ctx) {
        return visit(ctx.expr());
    }

    @Override
    public Double visitInt(LabeledExprParser.IntContext ctx) {
        return Double.valueOf(ctx.INT().getText());
    }

    @Override
    public Double visitId(LabeledExprParser.IdContext ctx) {
        String id = ctx.ID().getText();
        return memory.getOrDefault(id, 0.0);
    }

    @Override
    public Double visitParens(LabeledExprParser.ParensContext ctx) {
        return visit(ctx.expr());
    }

    @Override
    public Double visitMulDiv(LabeledExprParser.MulDivContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));

        return switch (ctx.op.getType()) {
            case LabeledExprParser.MUL -> left * right;
            case LabeledExprParser.DIV -> left / right;
            case LabeledExprParser.MOD -> left % right;
            case LabeledExprParser.POW -> Math.pow(left, right);
            default -> 0.0;
        };
    }

    @Override
    public Double visitAddSub(LabeledExprParser.AddSubContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));

        return switch (ctx.op.getType()) {
            case LabeledExprParser.ADD -> left + right;
            case LabeledExprParser.SUB -> left - right;
            default -> 0.0;
        };
    }

    @Override
    public Double visitFuncCall(LabeledExprParser.FuncCallContext ctx) {
        String funcName = ctx.func.getText().toLowerCase();
        double value = visit(ctx.expr());

        return switch (funcName) {
            case "sin" -> Math.sin(Math.toRadians(value));
            case "cos" -> Math.cos(Math.toRadians(value));
            case "tan" -> Math.tan(Math.toRadians(value));
            default -> {
                System.err.println("Funcion desconocida: " + funcName);
                yield 0.0;
            }
        };
    }
    @Override
    public Double visitFuncRad(LabeledExprParser.FuncRadContext ctx) {
        String funcName = ctx.func.getText().toLowerCase();
        double value = visit(ctx.expr());

        return switch (funcName) {
            case "sin" -> Math.sin(value);
            case "cos" -> Math.cos(value);
            case "tan" -> Math.tan(value);
            default -> {
                System.err.println("Funcion desconocida: " + funcName);
                yield 0.0;
            }
        };
    }
    @Override
    public Double visitLnf(LabeledExprParser.LnfContext ctx) {
        double value = visit(ctx.expr());
        if (value <= 0) {
            System.err.println("Error: logaritmo no definido para " + value);
            return Double.NaN;
        }
        return Math.log(value);
    }
    @Override
    public Double visitLogf(LabeledExprParser.LogfContext ctx) {
        double value = visit(ctx.expr());
        if (value <= 0) {
            System.err.println("Error: logaritmo no definido para " + value);return Double.NaN;
            
        }
        return Math.log10(value);
    }
    @Override
    public Double visitSqrtf(LabeledExprParser.SqrtfContext ctx) {
        double value = visit(ctx.expr());
        if (value < 0) {
            System.err.println("Error: raiz cuadrada no definida para " + value);
            return Double.NaN;
        }
        return Math.sqrt(value);
    }
    @Override
    public Double visitFactf(LabeledExprParser.FactfContext ctx) {
        double value = visit(ctx.expr());
        if (value < 0) {
            System.err.println("Error: factorial no definido para " + value);
            return Double.NaN;
        }
        return factorial(value);
    }

    private double factorial(double n) {
        if (n == 0) return 1;
        return n * factorial(n - 1);
    }
}

File: Python/Visitor.py
from LabeledExprVisitor import LabeledExprVisitor
from LabeledExprParser import LabeledExprParser
import math

class EvalVisitor(LabeledExprVisitor):
    def __init__(self):
        self.memory = {}

    def visitAssign(self, ctx:LabeledExprParser.AssignContext):
        id = ctx.ID().getText()
        value = self.visit(ctx.expr())
        self.memory[id] = value
        return value

    def visitPrintExpr(self, ctx:LabeledExprParser.PrintExprContext):
        value = self.visit(ctx.expr())
        print(value)
        return value

    def visitBlank(self, ctx:LabeledExprParser.BlankContext):
        return None


    def visitInt(self, ctx:LabeledExprParser.IntContext):
        return int(ctx.INT().getText())

    def visitDouble(self, ctx:LabeledExprParser.DoubleContext):
        return float(ctx.DOUBLE().getText())

    def visitId(self, ctx:LabeledExprParser.IdContext):
        id = ctx.ID().getText()
        return self.memory.get(id, 0)

    def visitParens(self, ctx:LabeledExprParser.ParensContext):
        return self.visit(ctx.expr())
    
    def visitUnaryMinus(self, ctx:LabeledExprParser.UnaryMinusContext):
        return -self.visit(ctx.expr())

    def visitUnaryPlus(self, ctx:LabeledExprParser.UnaryPlusContext):
        return +self.visit(ctx.expr())

    def visitAddSub(self, ctx:LabeledExprParser.AddSubContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == LabeledExprParser.ADD:
            return left + right
        return left - right

    def visitMulDiv(self, ctx:LabeledExprParser.MulDivContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == LabeledExprParser.MUL:
            return left * right
        elif ctx.op.type == LabeledExprParser.DIV:
            if right==0:
                print("Error: División por cero")
                return None
            return left / right
        elif ctx.op.type == LabeledExprParser.MOD:
            return left % right
        elif ctx.op.type == LabeledExprParser.POW:
            return left ** right
        else:
            return None

    def visitFactf(self, ctx:LabeledExprParser.FactfContext):
        value = self.visit(ctx.expr())
        return math.factorial(int(value))

    def visitFuncCall(self, ctx:LabeledExprParser.FuncCallContext):
        func = ctx.func.text.lower()
        value = self.visit(ctx.expr())
        if func == "sin":
            return math.sin(math.radians(value))
        elif func == "cos":
            return math.cos(math.radians(value))
        elif func == "tan":
            return math.tan(math.radians(value))
        else:
            print("Función desconocida:", func)
            return None

    def visitFuncRad(self, ctx:LabeledExprParser.FuncRadContext):
        func = ctx.func.text.lower()
        value = self.visit(ctx.expr())
        if func == "sin":
            return math.sin(value)   
        elif func == "cos":
            return math.cos(value)
        elif func == "tan":
            return math.tan(value)
        else:
            print("Función desconocida con rad:", func)
            return None

    def visitSqrtf(self, ctx:LabeledExprParser.SqrtfContext):
        if self.visit(ctx.expr()) < 0:
            print("Error: Raíz cuadrada de un número negativo")
            return None
            
        return math.sqrt(self.visit(ctx.expr()))

    def visitLnf(self, ctx:LabeledExprParser.LnfContext):
        if self.visit(ctx.expr()) <= 0:
            print("Error: Logaritmo natural de un número no positivo")
            return None
        return math.log(self.visit(ctx.expr()))

    def visitLogf(self, ctx:LabeledExprParser.LogfContext):
        if self.visit(ctx.expr()) <= 0:
            print("Error: Logaritmo de un número no positivo")
            return None
        return math.log10(self.visit(ctx.expr()))

    def visitProg(self, ctx:LabeledExprParser.ProgContext):
        result = None
        for child in ctx.getChildren():
            result = self.visit(child)
        return result
